Bugzilla/Install/AssetManager.pm

```diff
@@ -0,0 +1,218 @@
+        my $asset_file = File::Spec->catfile( $asset_dir, "$digest.$ext" );
+        cp( $file, $asset_file ) or die "failed to copy $file to $asset_file: $!";
+        if ( $digest eq $self->_digest_file($asset_file) ) {
+            $asset_map->{$key} = File::Spec->abs2rel( $asset_file, $asset_dir );
+        }
+        else {
+            die "failed to write $asset_file";
+        }
+    }
+    elsif ( $file =~ /\.css$/ ) {
+        my $content = read_file($file);
+
+        $content =~ s{(?<!=)url\(([^\)]+)\)}{$self->_css_url_rewrite($1, $file)}eig;
+        my $digest = $self->_digest_string($content);
+        my $asset_file = File::Spec->catfile( $asset_dir, "$digest.css" );
+        write_file( $asset_file, $content ) or die "failed to write $asset_file: $!";
+        if ( $digest eq $self->_digest_file($asset_file) ) {
+            $asset_map->{$key} = File::Spec->abs2rel( $asset_file, $asset_dir );
+        }
+        else {
+            die "failed to write $asset_file";
+        }
+    }
+}
+
+sub compile_all {
+    my ($self) = @_;
+    my $asset_map = $self->state->{asset_map} = {};
+
+    my $wanted = sub {
+        $self->compile_file($File::Find::name);
+    };
+
+    find( { wanted => $wanted, no_chdir => 1 }, @{ $self->source_dirs } );
+
+    $self->_save_state();
+}
+
+sub _css_url_rewrite {
+    my ( $self, $url, $file ) = @_;
+    my $dir = dirname($file);
+
+    # rewrite relative urls as the unified stylesheet lives in a different
+    # directory from the source
+    $url =~ s/(^['"]|['"]$)//g;
+    if ( $url =~ m!^(/|data:)! ) {
+        return 'url(' . $url . ')';
+    }
+    else {
+        my $url_file = File::Spec->rel2abs( $url, $dir );
+        my $ref_file = File::Spec->abs2rel( $url_file, $self->base_dir );
+        $self->compile_file($url_file);
+        return sprintf( "url(%s)", $self->asset_file( $ref_file, $self->asset_dir ) );
+    }
+}
+
+sub _new_digest { Digest::SHA->new($SHA_VERSION) }
+
+sub _digest_file {
+    my ( $self, $file ) = @_;
+    my $digest = $self->_new_digest;
+    $digest->addfile( $file, "b" );
+    return $digest->hexdigest;
+}
+
+sub _digest_string {
+    my ( $self, $string ) = @_;
+    my $digest = $self->_new_digest;
+    $digest->add($string);
+    return $digest->hexdigest;
+}
+
+sub _build_base_dir {
+    Cwd::realpath( File::Spec->rel2abs( bz_locations->{cgi_path} ) );
+}
+
+sub _build_asset_dir {
+    my ($self) = @_;
+    my $dir
+        = Cwd::realpath( File::Spec->rel2abs( bz_locations->{assetsdir} ) );
+
+    if ( $dir && -d $dir ) {
+        my $version_dir = File::Spec->catdir( $dir, "v" . $self->VERSION );
+        unless ( -d $version_dir ) {
+            mkdir $version_dir or die "mkdir $version_dir failed: $!";
+        }
+        return $version_dir;
+    }
+    else {
+        return $dir;
+    }
+}
+
+sub _build_source_dirs {
+    my ($self)  = @_;
+    my $base    = $self->base_dir;
+    my $ext_dir = "$base/extensions";
```

Comment:
I would encourage a blank line here, followed by ```my @dirs;```, so that it's apparent what the next three lines of (relatively dense) code are for, distinct from $base and returnvalue stuff.